src/bfcli/print.c

```diff
@@ -0,0 +1,167 @@
+/* SPDX-License-Identifier: GPL-2.0-only */
+/*
+ * Copyright (c) 2023 Meta Platforms, Inc. and affiliates.
+ */
+
+#include "print.h"
+
+#include <errno.h>
+#include <stdbool.h>
+#include <stddef.h>
+#include <stdio.h>
+
+#include "core/chain.h"
+#include "core/counter.h"
+#include "core/helper.h"
+#include "core/hook.h"
+#include "core/list.h"
+#include "core/logger.h"
+#include "core/matcher.h"
+#include "core/rule.h"
+#include "core/verdict.h"
+
+#define BF_DUMP_HEXDUMP_LEN 8
+#define BF_DUMP_TOKEN_LEN 5
+
+/**
+ * Dump a block of memory in hexadecimal format.
+ *
+ * @param data Pointer to the data to be dumped. Must be non-NULL.
+ * @param len Length of the data in bytes.
+ */
+static void bf_dump_hex_local(const void *data, size_t len)
+{
+    bf_assert(data);
+    // 5 characters per byte (0x%02x) + 1 for the null terminator.
+    char buf[(BF_DUMP_HEXDUMP_LEN * BF_DUMP_TOKEN_LEN) + 1];
+    const void *end = data + len;
+
+    while (data < end) {
+        char *line = buf;
+        for (size_t i = 0; i < BF_DUMP_HEXDUMP_LEN && data < end; ++i, ++data)
+            line += sprintf(line, "0x%02x ", *(unsigned char *)data);
+
+        (void)fprintf(stderr, "%s", buf);
+    }
+}
+
+/**
+ * Dump the details of a chain, including its rules and counters.
+ *
+ * @param chain Pointer to the chain to be dumped. Must be non-NULL.
+ * @param counter Pointer to the array of counters associated with the
+ *        chain. Must be non-NULL if with_counters is true.
+ * @param with_counters Boolean flag indicating whether to include
+ *        counters in the dump.
+ */
+static int bf_cli_chain_dump(struct bf_chain *chain, bf_list *counters,
+                             bool with_counters)
+{
+    struct bf_hook_opts *opts = &chain->hook_opts;
+    struct bf_counter *counter = NULL;
+
+    bf_assert(chain);
+    bf_assert(!with_counters || counters);
+
+    (void)fprintf(stderr, "chain %s", bf_hook_to_str(chain->hook));
+    (void)fprintf(stderr, "{");
+
+    (void)fprintf(stderr, "attach=%s,", opts->attach ? "yes" : "no");
+    (void)fprintf(stderr, "ifindex=%d", opts->ifindex);
+    if (opts->name)
+        (void)fprintf(stderr, ",name=%s", opts->name);
+    (void)fprintf(stderr, "}");
+    (void)fprintf(stderr, " policy: %s\n", bf_verdict_to_str(chain->policy));
+
+    if (with_counters) {
+        /**
+         * Rule counters are followed by policy and error counters.
+         * These bf_list_get_at() calls cost linear time.
+        */
+        counter = (struct bf_counter *)bf_list_get_at(
```

Comment:
Instead of using `bf_list_get_at()`, use `bf_list_get_head()` once and then `bf_list_node_next()`.